src/cdk-experimental/popover-edit/table-directives.ts

```diff
@@ -0,0 +1,291 @@
+  Input,
+  NgZone,
+  OnDestroy,
+  TemplateRef,
+  ViewContainerRef
+} from '@angular/core';
+import {fromEvent, ReplaySubject} from 'rxjs';
+import {debounceTime, filter, map, mapTo, takeUntil} from 'rxjs/operators';
+
+import {CELL_SELECTOR, EDIT_PANE_CLASS, EDIT_PANE_SELECTOR, ROW_SELECTOR} from './constants';
+import {EditEventDispatcher} from './edit-event-dispatcher';
+import {closest} from './polyfill';
+import {PopoverEditPositionStrategyFactory} from './popover-edit-position-strategy-factory';
+
+/**
+ * The delay between the mouse entering a row and the mouse stopping its movement before
+ * showing on-hover content.
+ */
+const DEFAULT_MOUSE_MOVE_DELAY_MS = 30;
+
+/**
+ * A directive that must be attached to enable editability on a table.
+ * It is responsible for setting up delegated event handlers and providing the
+ * EditEventDispatcher service for use by the other edit directives.
+ */
+@Directive({
+  selector: 'table[editable], cdk-table[editable], mat-table[editable]',
+  providers: [EditEventDispatcher],
+})
+export class CdkEditable implements AfterViewInit, OnDestroy {
+  protected readonly destroyed = new ReplaySubject<void>();
+
+  constructor(
+      protected readonly elementRef: ElementRef,
+      protected readonly editEventDispatcher: EditEventDispatcher,
+      protected readonly ngZone: NgZone) {}
+
+  ngAfterViewInit(): void {
+    this._listenForTableEvents();
+  }
+
+  ngOnDestroy(): void {
+    this.destroyed.next();
+    this.destroyed.complete();
+  }
+
+  private _listenForTableEvents(): void {
+    const element = this.elementRef.nativeElement!;
+
+    const toClosest = (selector: string) =>
+        map((event: UIEvent) => closest(event.target, selector));
+
+    this.ngZone.runOutsideAngular(() => {
+      fromEvent<MouseEvent>(element, 'mouseover').pipe(
+          takeUntil(this.destroyed),
+          toClosest(ROW_SELECTOR),
+          ).subscribe(this.editEventDispatcher.hovering);
+      fromEvent<MouseEvent>(element, 'mouseleave').pipe(
+          takeUntil(this.destroyed),
+          mapTo(null),
+          ).subscribe(this.editEventDispatcher.hovering);
+      fromEvent<MouseEvent>(element, 'mousemove').pipe(
+          takeUntil(this.destroyed),
+          debounceTime(DEFAULT_MOUSE_MOVE_DELAY_MS),
+          toClosest(ROW_SELECTOR),
+          ).subscribe(this.editEventDispatcher.mouseMove);
+
+      fromEvent<KeyboardEvent>(element, 'keyup').pipe(
+          takeUntil(this.destroyed),
+          filter(event => event.key === 'Enter'),
+          toClosest(CELL_SELECTOR),
+          ).subscribe(this.editEventDispatcher.editing);
+    });
+  }
+}
+
+/**
+ * Attaches an ng-template to a cell and shows it when instructed to by the
+ * EditEventDispatcher service.
+ * Makes the cell focusable.
+ */
+@Directive({
+  selector: '[cdkPopoverEdit]',
+  host: {
+    'tabIndex': '0',
+    '[attr.aria-haspopup]': 'true',
+  }
+})
+export class CdkPopoverEdit<Context> implements AfterViewInit, OnDestroy {
```

Comment:
Just `T` or `C` for the generic? This confused me for a minute trying to figure out what `Context` was